python/pyarrow/_cuda.pyx

```diff
@@ -0,0 +1,760 @@
+
+    Returns
+    -------
+    message : Message
+      the deserialized message, body still on device
+    """
+    cdef:
+        Message result = Message.__new__(Message)
+    cdef CMemoryPool* pool_ = maybe_unbox_memory_pool(pool)
+    if not isinstance(source, BufferReader):
+        reader = BufferReader(source)
+    check_status(CudaReadMessage(reader.reader, pool_, &result.message))
+    return result
+
+
+def read_record_batch(object schema, object buffer, pool=None):
+    """Construct RecordBatch referencing IPC message located on CUDA device.
+
+    While the metadata must be copied to host memory for
+    deserialization, the record batch data remains on the device.
+
+    Parameters
+    ----------
+    schema : Schema
+      Specify schema for the record batch
+    buffer :
+      Specify device buffer containing the complete IPC message
+    pool : {MemoryPool, None}
+      Specify pool to use for allocating space for the metadata
+
+    Returns
+    -------
+    batch : RecordBatch
+      reconstructed record batch, with device pointers
+
+    """
+    cdef shared_ptr[CSchema] schema_ = pyarrow_unwrap_schema(schema)
+    cdef shared_ptr[CCudaBuffer] buffer_ = pyarrow_unwrap_cudabuffer(buffer)
+    cdef CMemoryPool* pool_ = maybe_unbox_memory_pool(pool)
+    cdef shared_ptr[CRecordBatch] batch
+    check_status(CudaReadRecordBatch(schema_, buffer_, pool_, &batch))
+    return pyarrow_wrap_batch(batch)
+
+
+# Public API
+
+
+cdef public api bint pyarrow_is_buffer(object buffer):
+    return isinstance(buffer, Buffer)
+
+cdef public api bint pyarrow_is_cudabuffer(object buffer):
+    return isinstance(buffer, CudaBuffer)
+
+cdef public api bint pyarrow_is_cudahostbuffer(object buffer):
+    return isinstance(buffer, HostBuffer)
+
+cdef public api bint pyarrow_is_cudacontext(object ctx):
+    return isinstance(ctx, Context)
+
+cdef public api object \
+        pyarrow_wrap_cudabuffer(const shared_ptr[CCudaBuffer] &buf):
+    cdef CudaBuffer result = CudaBuffer.__new__(CudaBuffer)
+    result.init_cuda(buf)
+    return result
+
+cdef public api object \
+        pyarrow_wrap_cudahostbuffer(const shared_ptr[CCudaHostBuffer] &buf):
+    cdef HostBuffer result = HostBuffer.__new__(HostBuffer)
+    result.init_host(buf)
+    return result
+
+cdef public api object \
+        pyarrow_wrap_cudacontext(const shared_ptr[CCudaContext] &ctx):
+    cdef Context result = Context.__new__(Context)
+    result.init(ctx)
+    return result
+
+cdef public api bint pyarrow_is_cudaipcmemhandle(object handle):
+    return isinstance(handle, IpcMemHandle)
+
+cdef public api object \
+        pyarrow_wrap_cudaipcmemhandle(shared_ptr[CCudaIpcMemHandle] &h):
+    cdef IpcMemHandle result = IpcMemHandle.__new__(IpcMemHandle)
+    result.init(h)
+    return result
+
+cdef public api shared_ptr[CCudaIpcMemHandle] \
+        pyarrow_unwrap_cudaipcmemhandle(object handle):
+    cdef IpcMemHandle handle_
+    assert isinstance(handle, IpcMemHandle)
```

Comment:
Using more uniform coding now.
While the public-api.pxi returns `shared_ptr[<C class name>]()` for objects that cannot be unwrapped to the required type, I choose to raise TypeError immediately instead of waiting for segfault conditions.